minidump-processor/src/op_analysis.rs

```diff
@@ -0,0 +1,733 @@
+///
+/// 2.  Calling `read()` one-or-more times with the stream position at an occupied address should
+///     continue to produce bytes until the end of the occupied memory range is reached, at which
+///     point `Ok(0)` should be returned at-least once to signal end-of-file. Specifically, it is
+///     never okay to "jump" across a gap to the next occupied range of memory, as that would have
+///     the effect of concatenating disjoint areas of memory
+///
+/// 3.  Using `seek()` to reposition the stream back to a occupied memory address should allow
+///     usage of `read()` to resume as normal
+///
+/// 4.  Both `seek()` and `read()` should return any I/O errors encountered
+///
+/// As long as those requirements are met, many other things are permitted. Specifically, all the
+/// following things are allowed but not required:
+///
+/// 1.  The stream may choose to fail if `seek(SeekFrom::Start(pos))` is not called before
+///     `read()`, as this module will always do that
+///
+/// 2.  The stream may support `SeekFrom::Current` or `SeekFrom::End`, but this module doesn't
+///     use either
+///
+/// 3.  After `read()` returns `Ok(0)` to indicate the end-of-file, the stream may return an error
+///     if `read()` is called again or may choose to return `Ok(0)`
+///
+/// 4.  A `seek()` to an unoccupied address may return an error, or may update the pointer and
+///     return success
+///
+/// 5.  Likewise, a `read()` of an unoccupied address may return an error, or may return `Ok(0)`
+///     to indicate that end-of-file is immediately hit
+///
+/// Implementing this marker trait for a type that doesn't meet this requirements has safe but
+/// unspecified behavior
+///
+/// As these requirements are fairly lenient, this means that any `File` or `&[u8]` that contains
+/// a flat memory dump of process memory may implement this trait
+///
+pub trait SparseMemoryStream: Read + Seek {}
+
+/// Read the instruction bytes that were being run by the given thread
+///
+/// Use the given `context` to attempt to read `1 <= n <= MAX_INSTRUCTION_LENGTH`
+/// bytes at the instruction pointer from the byte stream passed in `memory`, which is any type
+/// that implements `Seek + Read` and is meant to represent the memory of the target machine
+///
+///
+///
+/// # Errors
+///
+/// This may fail if the underlying byte stream fails to `seek()` or `read()`
+///
+pub fn get_thread_instruction_bytes(
+    context: &MinidumpContext,
+    memory: &mut impl SparseMemoryStream,
+) -> Result<Vec<u8>, OpAnalysisError> {
+    let instruction_pointer = context.get_instruction_pointer();
+
+    memory
+        .seek(SeekFrom::Start(instruction_pointer))
+        .map_err(OpAnalysisError::ReadThreadInstructionFailed)?;
+
+    // We use MAX_INSTRUCTION_LENGTH here as an optimization to avoid allocating and copying
+    // more bytes than needed to produce an instruction
+
+    let mut buffer = Vec::with_capacity(MAX_INSTRUCTION_LENGTH.into());
```

Comment:
Is it at all anticipated that MAX_INSTRUCTION_LENGTH can become absurdly large?

(I am mostly just curious if there's something fucked up on some architecture that you know about.)

Reply:
I'm not really aware of any architecture where an instruction would be very large. I guess it's theoretically possible these days with CPUs having so much onboard memory, but I doubt we'll be dealing with anything like that anytime soon!